internal/core/src/storage/MmapChunkManager.cpp

```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+MmapBlockPtr
+MmapBlocksHandler::AllocateLargeBlock(const uint64_t size) {
+    if (size + Capacity() > max_disk_limit_) {
+        ClearCache();
+    }
+    if (size + Size() > max_disk_limit_) {
+        PanicInfo(ErrorCode::MemAllocateSizeNotMatch,
+                  "Failed to create a new mmap_block, not enough disk for "
+                  "create a new mmap block. Allocated size: {}, Max size: {}",
+                  Size(),
+                  max_disk_limit_);
+    }
+    auto new_block = std::make_unique<MmapBlock>(
+        GetMmapFilePath(), size, MmapBlock::BlockType::Variable);
+    new_block->Init();
+    return std::move(new_block);
+}
+
+void
+MmapBlocksHandler::Deallocate(MmapBlockPtr&& block) {
+    if (block->GetType() == MmapBlock::BlockType::Fixed) {
+        // store the mmap block in cache
+        block->Reset();
+        fix_size_blocks_cache_.push(std::move(block));
+        uint64_t max_cache_size =
+            uint64_t(cache_threshold * (float)max_disk_limit_);
+        if (fix_size_blocks_cache_.size() * fix_mmap_file_size_ >
+            max_cache_size) {
+            FitCache(max_cache_size);
+        }
+    } else {
+        // release the mmap block
+        block->Close();
+        block = nullptr;
+    }
+}
+
+void
+MmapBlocksHandler::ClearCache() {
+    while (!fix_size_blocks_cache_.empty()) {
+        auto block = std::move(fix_size_blocks_cache_.front());
+        block->Close();
+        fix_size_blocks_cache_.pop();
+    }
+}
+
+void
+MmapBlocksHandler::FitCache(const uint64_t size) {
+    while (fix_size_blocks_cache_.size() * fix_mmap_file_size_ > size) {
+        auto block = std::move(fix_size_blocks_cache_.front());
+        block->Close();
+        fix_size_blocks_cache_.pop();
+    }
+}
+
+MmapChunkManager::~MmapChunkManager() {
+    // munmap all mmap_blocks before remove dir
+    for (auto it = blocks_table_.begin(); it != blocks_table_.end();) {
+        it = blocks_table_.erase(it);
+    }
+    if (blocks_handler_ != nullptr) {
+        blocks_handler_ = nullptr;
+    }
+    // clean the mmap dir
+    auto cm =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    if (cm->Exist(mmap_file_prefix_)) {
+        cm->RemoveDir(mmap_file_prefix_);
+    }
+}
+
+void
+MmapChunkManager::Register(const MmapChunkDescriptor key) {
+    if (HasKey(key)) {
+        LOG_WARN("key has exist in growing mmap manager");
+        return;
+    }
+    std::unique_lock<std::shared_mutex> lck(mtx_);
+    blocks_table_.emplace(key, std::vector<MmapBlockPtr>());
+    return;
+}
+
+void
+MmapChunkManager::UnRegister(const MmapChunkDescriptor key) {
+    std::unique_lock<std::shared_mutex> lck(mtx_);
+    if (blocks_table_.find(key) != blocks_table_.end()) {
+        auto& blocks = blocks_table_[key];
+        for (auto i = 0; i < blocks.size(); i++) {
+            blocks_handler_->Deallocate(std::move(blocks[i]));
+        }
+        blocks_table_.erase(key);
```

Comment:
use a folly concurrent hashmap and removed global lock.
Deallocated and allocated maybe time consuming so better aovid did it with holding the lock

Reply:
This step of optimization will be carried out after the relevant test results are stable.